### PR TITLE
True flyweight 2

### LINKED ISSUE #145: RDKit properties are not stored in `SmallMoleculeComponent.to_dict`
Problem illustrated below. However, there's a more fundamental issue that we were originally using RDKit properties to label the version of `gufe` that the molecule came from. So if we do include those in `to_dict`, `SmallMoleculeComponent` would need a custom gufe-key representation that excludes the gufe version properties from the hash.

My proposed solution would be to:

1. Include properties in `to_dict`
2. Remove the additional properties we add for gufe versioning.

Alternate solutions: 

* Custom `_tokenization` method for `SmallMoleculeComponent` that filters the specific properties we add 
* Drop the guarantee that we preserve properties attached to an rdkit molecule -- I'm not sure whether these properties would survive a round-trip to another molecular representation anyway (openff, oechem).

MCVE below (note that this has to be done in separate processes for the error to occur because our flyweight will reuse the same object in the same process).

```python
# dump_smc.py
import gufe
from gufe.tokenization import JSON_HANDLER
from rdkit import Chem
import json

mol = Chem.MolFromSmiles("CC")
Chem.AllChem.Compute2DCoords(mol)
mol.SetProp("foo", "bar")
smc = gufe.SmallMoleculeComponent.from_rdkit(mol)

assert smc.to_rdkit().GetProp("foo") == "bar"

dct = smc.to_dict()
with open("dump.json", mode='w') as f:
    f.write(json.dumps(dct, cls=JSON_HANDLER.encoder))
```

```python
# load_smc.py
import gufe
from gufe.tokenization import JSON_HANDLER
import json

with open("dump.json", mode='r') as f:
    dct = json.load(f)

smc = gufe.SmallMoleculeComponent.from_dict(dct)
assert smc.to_rdkit().GetProp("foo") == "bar"
```

```bash
$ python dump_smc.py
$ python load_smc.py
```

```pytb
Traceback (most recent call last):
  File "/Users/dwhs/Dropbox/omsf/tmp-rdkit-prop-issue/load_smc.py", line 9, in <module>
    assert smc.to_rdkit().GetProp("foo") == "bar"
KeyError: 'foo'
```

### PR DESCRIPTION
**Pull request by richardjgowers**
fixes #145

### CODE REVIEW

**Review by dwhswenson — CHANGES_REQUESTED** [review]
Important point here is that, by default, the `to_dict` serialization also defines the gufe key, which we would like to be as stable as possible. You can avoid this by overriding `_gufe_tokenize` method for this class.

We attach a property `ofe-version` to our molecules. Is that included in this serialization? (I think so.) If that is used to compute the gufe key, then the key is no longer stable against minor version updates (or even patch updates -- possibly even between commits?). I think the choices are (1) remove that property from the molecule altogether; (2) remove that property from the serialization (requiring something more custom); (3) write a custom `_gufe_tokenization` method that excludes that information.

I'd also like to know if we're sure we have a guarantee of stability in `RDMol.ToBinary`. Basing our tokenization on that means that changes there will break the stability of our gufe keys.

**Comment by richardjgowers** [comment]
@dwhswenson what's the point of ofe-version?  It sounds like we were keeping track of version for serialisation reasons, but maybe that's defunct and/or something we should be handling at the `GufeTokenizable` level?

**Comment by dwhswenson** [comment]
> @dwhswenson what's the point of ofe-version? It sounds like we were keeping track of version for serialisation reasons, but maybe that's defunct and/or something we should be handling at the `GufeTokenizable` level?

This is why I suggested removing it in #145 (point 2, "Remove the additional properties we add for gufe versioning")

**Comment by richardjgowers** [comment]
@dwhswenson I think pickles are intended to be stable, see here: https://github.com/rdkit/rdkit/blob/master/Code/GraphMol/MolPickler.h#L76

File: gufe/tests/test_smallmoleculecomponent.py [hunk 1]
```diff
@@ -83,13 +83,6 @@ class TestSmallMoleculeComponent(GufeTokenizableTestsMixin):
     def instance(self, named_ethane):
         return named_ethane
 
-    def test_rdkit_behavior(self, ethane, alt_ethane):
-        # Check that fixture setup is correct (we aren't accidentally
-        # testing tautologies)
-        assert ethane is not alt_ethane
-        assert ethane.to_rdkit() is not alt_ethane.to_rdkit()
-
-
```
[the remark this answers is not in the document]
Comment (dwhswenson) [reply gufe/tests/test_smallmoleculecomponent.py:92]: To be clear, we have `ethane is alt_ethane` now. And we test that `hash(ethane) == hash(alt_ethane)`.

This is the equivalent of testing `foo == foo`, not `foo == bar`.

**Pull request merged**